### PR TITLE
fix issue #4577 crash by ensuring the flag item's name is never null in rz_core_bin_apply_symbols

### LINKED ISSUE #4577: RISC-V 32: Crash when trying to display disassembly code
**Environment information**
* Operating System: MacOS 14.5 (ARM-64 - M1)
* Cutter version: 2.3.4-stable-209c26b
* Obtained from:
  - [ ] Built from source
  - [ ] Downloaded release from Cutter website or GitHub 
  - [x] Distribution repository (brew)
* File format:
ELF (RISC-V 32)

**Describe the bug**

Crash when trying to display disassembly code.

<details>
  <summary>MacOS report</summary>
 
```
-------------------------------------
Translated Report (Full Report Below)
-------------------------------------

Process:               cutter [75863]
Path:                  /Applications/Cutter.app/Contents/MacOS/cutter
Identifier:            re.rizin.cutter
Version:               2.3.4-stable-209c26b (2.3.4-stable-209c26b)
Code Type:             ARM-64 (Native)
Parent Process:        launchd [1]
User ID:               501

Date/Time:             2024-07-24 23:23:32.0016 +0400
OS Version:            macOS 14.5 (23F79)
Report Version:        12
Anonymous UUID:        5BC71BD9-F14A-2CB9-7C3A-A5EA18D4D253

Sleep/Wake UUID:       FF657A00-3D1F-4900-8568-BBFC9479AF65

Time Awake Since Boot: 1800000 seconds
Time Since Wake:       114638 seconds

System Integrity Protection: enabled

Crashed Thread:        0  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_BAD_ACCESS (SIGSEGV)
Exception Codes:       KERN_INVALID_ADDRESS at 0x0000000000000000
Exception Codes:       0x0000000000000001, 0x0000000000000000

Termination Reason:    Namespace SIGNAL, Code 11 Segmentation fault: 11
Terminating Process:   exc handler [75863]

VM Region Info: 0 is not in any region.  Bytes before following region: 4307992576
      REGION TYPE                    START - END         [ VSIZE] PRT/MAX SHRMOD  REGION DETAIL
      UNUSED SPACE AT START
--->  
      __TEXT                      100c6c000-100f84000    [ 3168K] r-x/r-x SM=COW  /Applications/Cutter.app/Contents/MacOS/cutter

Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
0   libsystem_platform.dylib      	       0x1848393c8 _platform_strstr + 32
1   librz_core.0.7.dylib          	       0x101fdd6cc ds_print_ptr + 2636
2   librz_core.0.7.dylib          	       0x101fd8a84 rz_core_print_disasm + 14536
3   cutter                        	       0x100ccf084 0x100c6c000 + 405636
4   cutter                        	       0x100cce59c 0x100c6c000 + 402844
5   cutter                        	       0x100cce1c0 0x100c6c000 + 401856
6   QtCore                        	       0x103d300ec 0x103b28000 + 2130156
7   cutter                        	       0x100c7bab8 CutterSeekable::seekableSeekChanged(unsigned long long, CutterCore::SeekHistoryType) + 72
8   QtCore                        	       0x103d300ec 0x103b28000 + 2130156
9   cutter                        	       0x100c7eaf0 CutterCore::seekChanged(unsigned long long, CutterCore::SeekHistoryType) + 72
10  cutter                        	       0x100c9abd8 CutterCore::seek(unsigned long long) + 140
11  cutter                        	       0x100c99ebc CutterCore::seekAndShow(unsigned long long) + 20
12  QtCore                        	       0x103d300ec 0x103b28000 + 2130156
13  QtWidgets                     	       0x10271d99c QAbstractItemView::activated(QModelIndex const&) + 52
14  QtWidgets                     	       0x10278a16c QTreeView::mouseDoubleClickEvent(QMouseEvent*) + 868
15  QtWidgets                     	       0x10250ea1c QWidget::event(QEvent*) + 128
16  QtWidgets                     	       0x1025a31e0 QFrame::event(QEvent*) + 56
17  QtWidgets                     	       0x10271c450 QAbstractItemView::viewportEvent(QEvent*) + 1124
18  QtWidgets                     	       0x102786cec QTreeView::viewportEvent(QEvent*) + 500
19  QtCore                        	       0x103d0015c QCoreApplicationPrivate::sendThroughObjectEventFilters(QObject*, QEvent*) + 264
20  QtWidgets                     	       0x1024d8d58 QApplicationPrivate::notify_helper(QObject*, QEvent*) + 260
21  QtWidgets                     	       0x1024db6a4 QApplication::notify(QObject*, QEvent*) + 6072
22  QtCore                        	       0x103cffe44 QCoreApplication::notifyInternal2(QObject*, QEvent*) + 208
23  QtWidgets                     	       0x1024d96dc QApplicationPrivate::sendMouseEvent(QWidget*, QMouseEvent*, QWidget*, QWidget*, QWidget**, QPointer<QWidget>&, bool, bool) + 968
24  QtWidgets                     	       0x10252c250 0x1024c8000 + 410192
25  QtWidgets                     	       0x10252b280 0x1024c8000 + 406144
26  QtWidgets                     	       0x1024d8d78 QApplicationPrivate::notify_helper(QObject*, QEvent*) + 292
27  QtWidgets                     	       0x1024da110 QApplication::notify(QObject*, QEvent*) + 548
28  QtCore                        	       0x103cffe44 QCoreApplication::notifyInternal2(QObject*, QEvent*) + 208
29  QtGui                         	       0x10300b23c QGuiApplicationPrivate::processMouseEvent(QWindowSystemInterfacePrivate::MouseEvent*) + 4436
30  QtGui                         	       0x102ff0558 QWindowSystemInterface::sendWindowSystemEvents(QFlags<QEventLoop::ProcessEventsFlag>) + 248
31  libqcocoa.dylib               	       0x101292d58 0x10125c000 + 224600
32  CoreFoundation                	       0x1848ea4d8 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 28
33  CoreFoundation                	       0x1848ea46c __CFRunLoopDoSource0 + 176
34  CoreFoundation                	       0x1848ea1dc __CFRunLoopDoSources0 + 244
35  CoreFoundation                	       0x1848e8dc8 __CFRunLoopRun + 828
36  CoreFoundation                	       0x1848e8434 CFRunLoopRunSpecific + 608
37  HIToolbox                     	       0x18f08c19c RunCurrentEventLoopInMode + 292
38  HIToolbox                     	       0x18f08be2c ReceiveNextEventCommon + 220
39  HIToolbox                     	       0x18f08bd30 _BlockUntilNextEventMatchingListInModeWithFilter + 76
40  AppKit                        	       0x188147d68 _DPSNextEvent + 660
41  AppKit                        	       0x18893d808 -[NSApplication(NSEventRouting) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 700
42  AppKit                        	       0x18813b09c -[NSApplication run] + 476
43  libqcocoa.dylib               	       0x101291b6c 0x10125c000 + 220012
44  QtCore                        	       0x103cfbf1c QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) + 524
45  QtCore                        	       0x103d00470 QCoreApplication::exec() + 132
46  cutter                        	       0x100c947c0 0x100c6c000 + 165824
47  dyld                          	       0x1844820e0 start + 2360
```
</details>

**To Reproduce**

1. Open file: `blink-crash.elf` from [blink.zip](https://github.com/user-attachments/files/16366938/blink.zip)

2. Try open function `blink.copy_data`
3. \<Crash>

**Expected behavior**

Show disassembly code.

**Additional context**

If I remove the `copy_data` function, then the crash does not occur(file `blink-no-copy_data.elf` in archive).

### PR DESCRIPTION
**Pull request by moste00**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [x] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

This is a quick non-root-cause fix for https://github.com/rizinorg/rizin/issues/4577. Briefly, the issue is that cutter crashes on an elf binary which contains symbols with no name (which are legal in elf files) when clicking on a function's name after opening the file. 

After some investigation, it turns out that the function `ds_print_ptr` in `disasm.c ` is trying to call `strstr` on the `realname` member of a `FlagItem` instance representing 2 offsets but the field is `NULL`. 

It's not clear why the other thousands of offsets/symbols with `FlagItem` instances having a `NULL` `realname` don't crash `ds_print_ptr` similarly. It's also not clear if this issue generalizes beyond the specific arch/compiler combination of the crash-inducing binary (RISC-V and the Zig compiler, respectively), it doesn't appear to have anything to do with RISC-V or Zig, but further investigation is required to pin down the exact reason of the crash on those particular `FlagItems`.

**Test plan**

I tried this fix in the rizin sub-directory of cutter and tried to reproduce the crash, crash no longer reproduces. 

**Closing issues**

closes #4577.

### CODE REVIEW

**Review by Rot127 — APPROVED** [review]

**Comment by Rot127** [comment]
The two failing tests seem to be related. Maybe some issue with the dwarf info returning NULL somewhere?

File: librz/core/cbin.c [hunk 1]
```diff
@@ -1563,7 +1563,7 @@ RZ_API bool rz_core_bin_apply_symbols(RzCore *core, RzBinFile *binfile, bool va)
 				}
 
 				fi = rz_flag_set(core->flags, fnp, addr, symbol->size);
-				if (fi) {
+				if (fi && !RZ_STR_ISEMPTY(n)) {
```
Comment (wargio) [inline librz/core/cbin.c:1566]: According to the code `n` is `const char *n = sn.demname ? sn.demname : symbol->name;`.
`n` to be `NULL` must have `symbol->name` as `NULL`.

This is weird, can you share the binary that does this?

File: librz/core/cbin.c [hunk 2]
```diff
@@ -1563,7 +1563,7 @@ RZ_API bool rz_core_bin_apply_symbols(RzCore *core, RzBinFile *binfile, bool va)
...
 					rz_flag_item_set_realname(fi, n);
```
Comment (wargio) [inline librz/core/cbin.c:1567]: Can you add this test before where if `symbol->name` is empty we then use `rz_str_newf("unk_%08" PFMT64x", offset);`

Comment (wargio) [reply librz/core/cbin.c:1567]: This would output something more meaningful in the failing test.

```patch
-- stdout
--- expected
+++ actual
@@ -30,8 +30,7 @@
 |       ,=< 0x0000803e      cbnz  r3, 0x804e
 |       |   0x00008040      ldr   r3, [data.00008054]                  ; [0x8054:4]=0
 |      ,==< 0x00008042      cbz   r3, 0x804a
-|      ||   0x00008044      ldr   r0,                                  ; data.00008058
-|      ||                                                              ; [0x8058:4]=0x977c obj.__FRAME_END
+|      ||   0x00008044      ldr   r0, obj.                             ; [data.00008058:4]=0x977c obj.__FRAME_END
 |      ||   0x00008046      nop.w
 |      `--> 0x0000804a      movs  r3, 1
 |       |   0x0000804c      strb  r3, [r4]
```

```
-- stderr
WARNING: core: cannot set flag with name 'obj.'
```

Comment (moste00) [reply librz/core/cbin.c:1567]: Ok, this is done locally, but there is some weird behaviour. When I write 

```
if (RZ_STR_ISEMPTY(n)) {
	n = rz_str_newf("unk_%08" PFMT64x, addr);
}
```

The output is indeed something like 

` 0x00008044      ldr   r0, [unk_0000977c]                   ; [data.00008058:4]=0x977c obj.__FRAME_END`

But I just add `0x` before the address (after `"unk_`), the output is now suddenly

`0x00008044      ldr   r0, [unk_obj.__FRAME_END]            ; [data.00008058:4]=0x977c obj.__FRAME_END`

Is there some search-and-replace pass that replaces all addresses with their names? Where did that name come from and why is it not be the original name to begin with (`symbol->name`)? Could we exploit this and just remove the "unk_" prefix and let the printed symbol be either [name] or [0xaddr]?

Comment (Rot127) [reply librz/core/cbin.c:1567]: Just guessing. But it probably is the function `find_next_number(char *str)` then. It is used to replace addresses and such with symbol names. Numbers like `0x8912` etc are replaced (see in-code comment about the regex).

I think it is best to just add the `unk_` prefix. Don't touch the filtering/replacement.

This whole `filter.c` code is one of the worst places in Rizin. It will be replaced at some point. See: https://github.com/rizinorg/rizin/issues/4238.

Comment (moste00) [reply librz/core/cbin.c:1567]: Right, but which would be better in your view, unk_number or unk_obj.name ?

Comment (Rot127) [reply librz/core/cbin.c:1567]: `unk_obj.name` I think. Because the number/address is still visible in the comment on the right.

Comment (wargio) [reply librz/core/cbin.c:1567]: Yes. This should be fixed on elf side. Just add it without the 0x

Comment (wargio) [reply librz/core/cbin.c:1567]: What rot said is correct, but I decided that this fix is still not ok. we should fix the root cause not the code in between.
This fix (https://github.com/rizinorg/rizin/pull/4821) should patch the issue

**Comment by wargio** [comment]
Closing this since the root cause has been patched. https://github.com/rizinorg/rizin/pull/4821

@moste00 thanks anyway for this patch. :)
